File: tests/fixtures/sample_repo/rust_pkg/src/calculator.rs
//! Stateful calculator with history tracking.

use crate::models::{CalculationRecord, Operation};

/// Error returned when division by zero is attempted.
#[derive(Debug, thiserror::Error)]
pub enum CalculatorError {
    #[error("division by zero")]
    DivisionByZero,
}

/// A stateful calculator that records all performed operations.
pub struct Calculator {
    history: Vec<CalculationRecord>,
}

impl Calculator {
    /// Creates a new, empty `Calculator`.
    pub fn new() -> Self {
        Self { history: Vec::new() }
    }

    /// Adds `x` and `y`, records the result, and returns it.
    pub fn add(&mut self, x: f64, y: f64) -> f64 {
        let result = x + y;
        self.record(Operation::Add, x, y, result);
        result
    }

    /// Subtracts `y` from `x`, records the result, and returns it.
    pub fn subtract(&mut self, x: f64, y: f64) -> f64 {
        let result = x - y;
        self.record(Operation::Subtract, x, y, result);
        result
    }

    /// Multiplies `x` by `y`, records the result, and returns it.
    pub fn multiply(&mut self, x: f64, y: f64) -> f64 {
        let result = x * y;
        self.record(Operation::Multiply, x, y, result);
        result
    }

    /// Divides `x` by `y`. Returns [`CalculatorError::DivisionByZero`] if `y == 0`.
    pub fn divide(&mut self, x: f64, y: f64) -> Result<f64, CalculatorError> {
        if y == 0.0 {
            return Err(CalculatorError::DivisionByZero);
        }
        let result = x / y;
        self.record(Operation::Divide, x, y, result);
        Ok(result)
    }

    /// Returns a slice of all recorded calculations.
    pub fn history(&self) -> &[CalculationRecord] {
        &self.history
    }

    /// Clears the history.
    pub fn clear_history(&mut self) {
        self.history.clear();
    }

    fn record(&mut self, op: Operation, x: f64, y: f64, result: f64) {
        self.history.push(CalculationRecord::new(op, x, y, result));
    }
}

impl Default for Calculator {
    fn default() -> Self {
        Self::new()
    }
}

File: tests/fixtures/sample_repo/go_pkg/calculator/calculator.go
// Package calculator provides stateful arithmetic operations with history.
package calculator

import (
	"errors"
	"fmt"

	"github.com/repowise-ai/sample/types"
)

// ErrDivisionByZero is returned when a division-by-zero is attempted.
var ErrDivisionByZero = errors.New("division by zero")

// Calculator performs arithmetic operations and records their history.
type Calculator struct {
	history []types.CalculationRecord
}

// New returns a new, empty Calculator.
func New() *Calculator {
	return &Calculator{}
}

// Add returns the sum of the operands and appends the result to history.
func (c *Calculator) Add(ops types.Operands) (float64, error) {
	result := ops.X + ops.Y
	c.record(types.OpAdd, ops, result)
	return result, nil
}

// Subtract returns ops.X - ops.Y and appends the result to history.
func (c *Calculator) Subtract(ops types.Operands) (float64, error) {
	result := ops.X - ops.Y
	c.record(types.OpSubtract, ops, result)
	return result, nil
}

// Multiply returns the product of the operands.
func (c *Calculator) Multiply(ops types.Operands) (float64, error) {
	result := ops.X * ops.Y
	c.record(types.OpMultiply, ops, result)
	return result, nil
}

// Divide returns ops.X / ops.Y. Returns ErrDivisionByZero if ops.Y == 0.
func (c *Calculator) Divide(ops types.Operands) (float64, error) {
	if ops.Y == 0 {
		return 0, ErrDivisionByZero
	}
	result := ops.X / ops.Y
	c.record(types.OpDivide, ops, result)
	return result, nil
}

// History returns a copy of all calculation records.
func (c *Calculator) History() []types.CalculationRecord {
	out := make([]types.CalculationRecord, len(c.history))
	copy(out, c.history)
	return out
}

// ClearHistory removes all recorded calculations.
func (c *Calculator) ClearHistory() {
	c.history = c.history[:0]
}

// record appends a new entry to the internal history slice.
func (c *Calculator) record(op types.Operation, ops types.Operands, result float64) {
	c.history = append(c.history, types.CalculationRecord{
		Operation: op,
		Operands:  ops,
		Result:    result,
		Summary:   fmt.Sprintf("%.2f %s %.2f = %.2f", ops.X, op, ops.Y, result),
	})
}
